x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java

```diff
@@ -311,9 +313,12 @@ public synchronized void setShouldStopAtCheckpoint(
             return;
         }
 
-        // move the call to the generic thread pool, so we do not block the network thread
-        getThreadPool().executor(ThreadPool.Names.GENERIC)
```

Comment:
this is the origin of the deadlock, by moving to another thread we unlocked the task (this method is `synchronized`).